### PR TITLE
Disable stylesheet objects rather than link elements

### LINKED ISSUE #75011: rustdoc page styles are broken with Waterfox
https://docs.rs does not work with Waterfox because it relies on the non-standard `disabled=""` Attribute in the theme `<link>` tags:

```html
      <link href="/light-20200726-1.47.0-nightly-6c8927b0c.css" id="themeStyle" rel="stylesheet" type="text/css">
      <link disabled="" href="/dark-20200726-1.47.0-nightly-6c8927b0c.css" rel="stylesheet" type="text/css">
      <link disabled="" href="/ayu-20200726-1.47.0-nightly-6c8927b0c.css" rel="stylesheet" type="text/css">
```

Without "disabled" support (which is the case with Waterfox), all three links will be enabled and will make the (imo unreadable) "ayu" the active one.

Accordingly https://developer.mozilla.org/en-US/docs/Web/HTML/Element/link this attribute requires very recent browsers and is marked as "Deprecated: do not use for new websites"

### PR DESCRIPTION
**Pull request by Cldfire**
Resolves #75011.

I changed the way themes are switched in rustdoc to more closely resemble [that of mdbook](https://github.com/rust-lang/mdBook/blob/a884c2574e5729a8c0e25e3c5c4302edfba11485/src/theme/book.js#L307-L326). This method uses [StyleSheet.disabled](https://developer.mozilla.org/en-US/docs/Web/API/StyleSheet/disabled) which has much better browser support than the usage of `disabled` on `link` that I introduced in a previous PR.

As added bonuses, we no longer need to use the IDs `mainThemeStyle` or `themeStyle`, usage of `switchTheme` is simpler, and we no longer need the hack that changes the href on the `light.css` stylesheet.

I've tested this in Firefox, Chromium, and Waterfox Classic, and it works well across all those browsers. @Nemo157 would perhaps be able to test this in Safari as well?

r? @GuillaumeGomez

### CODE REVIEW

File: src/librustdoc/html/static/storage.js
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/storage.js:114]:
```suggestion
themeSheet.disabled = themeName !== newTheme;
```

**Review by GuillaumeGomez — CHANGES_REQUESTED** [review]
I'm worried about this change: until the `storage.js` file is executed, all themes are enabled at the same time, right? This is a huge step backward. :-/

**Comment by Cldfire** [comment]
> I'm worried about this change: until the `storage.js` file is executed, all themes are enabled at the same time, right? This is a huge step backward. :-/

Correct, do you think that will be a problem?

**Comment by GuillaumeGomez** [comment]
On slowest computers, it's gonna be an issue for sure: the browser will have to parse and render multiple styles at once. Definitely not optimal...

**Comment by Cldfire** [comment]
> 
> 
> On slowest computers, it's gonna be an issue for sure: the browser will have to parse and render multiple styles at once. Definitely not optimal...

True, although this was the behavior of rustdoc before https://github.com/rust-lang/rust/commit/03f565cdbfbdda55798f2c6e1fd0a96ac1c35c76. As far as I'm aware this wasn't causing issues before with 2 theme stylesheets, do we believe it would cause an issue with 3?

The `disabled` attribute method on `link` elements is definitely better in this regard, though. If you don't see it becoming a compatibility issue it might be best to leave things as-is.

**Comment by GuillaumeGomez** [comment]
This is a good point. Only way would be to just compare I guess. Also, what do you @rust-lang/rustdoc ?

**Comment by jyn514** [comment]
Personally I don't think the themes are going to be affecting render time the most, it's the gobs of generated HTML. If you're worried though, maybe we could load the theme preference from local storage before loading the CSS?

**Comment by GuillaumeGomez** [comment]
> maybe we could load the theme preference from local storage before loading the CSS?

Isn't it already what we do?

**Comment by Manishearth** [comment]
> On slowest computers, it's gonna be an issue for sure: the browser will have to parse and render multiple styles at once. Definitely not optimal...


No, `<style disabled>` does not change parsing. What it does change is whether or not the style is applied, and flipping that attribute will cause a relayout to occur. If you set it in the `<head>` tag this will occur before any layout has happened so there should be no observable change.

**Comment by GuillaumeGomez** [comment]
@Manishearth This is what we do currently but this PR, unless I'm missing something, is removing this mechanism and the disable is added by the JS, not by rustdoc.

**Comment by Manishearth** [comment]
@GuillaumeGomez right, as long as the JS occurs in the head tag it doesn't matter


The parsing is inevitable. It's the relayout you want to avoid

**Comment by GuillaumeGomez** [comment]
Yes, on that point we agree. Thanks for the precision!

**Comment by JohnCSimon** [comment]
ping from triage: 
@Cldfire - can you address the change request and/or mark it as completed? Thanks.
https://github.com/rust-lang/rust/pull/75063#pullrequestreview-459671751

**Comment by jyn514** [comment]
This isn't waiting on anything AFAIK - @GuillaumeGomez are you happy with it?

**Comment by Cldfire** [comment]
Yep, this one is good to go if everyone else is happy :)

**Comment by GuillaumeGomez** [comment]
I guess we are. Moving forward then, thanks!

@bors: r+

**Comment by ollie27** [comment]
With JS disabled the ayu theme will now be used instead of the light theme. Is that change intentional?

**Comment by GuillaumeGomez** [comment]
Holding off until @ollie27's concern is resolved.

@bors: r-

**Comment by Cldfire** [comment]
> With JS disabled the ayu theme will now be used instead of the light theme. Is that change intentional?

With JS disabled the themes will now conflict with one another, so regardless of which stylesheet comes first in the HTML things could look funky.

Do we need to support no-JS usage? Doesn't that break search?

**Comment by jyn514** [comment]
Well, I think there's a difference between a feature that's JS-only and completely breaking the page if JS is turned off. I see the way this is implemented is `             themeSheet.disabled = themeName !== newTheme;` - is there a CSS equivalent of that? Or is that the `disabled=""` that isn't supported in WaterFox?

**Comment by jyn514** [comment]
Maybe we could have `light.css` for `<noscript>` and turn it on with JS only otherwise?

**Comment by GuillaumeGomez** [comment]
> Do we need to support no-JS usage? Doesn't that break search?

Yes and yes. We don't consider the search breakage when JS is disabled as an issue though, but the rendering has to be "clean", or at least readable.

**Comment by Manishearth** [comment]
Note: if this is broken under no-JS then it will also be broken if the relevant JS is in a separate file and needs to load, on bad networks. I think the default theme code is inline but if so we need a comment there that makes sure it's never moved out, and if not we need to make this work if that JS file fails to load

**Comment by Cldfire** [comment]
I investigated using [`rel="stylesheet alternate"`](https://developer.mozilla.org/en-US/docs/Web/CSS/Alternative_style_sheets), and while that does fix the no-JS side of things, it feels very laggy while switching, introduces a color flash while changing themes, and doesn't appear to have good browser support.

I came up with what appears to be a nice compromise here that still uses the `disabled` attribute on `link`. I changed it so that the `disabled` link attribute is used initially on page load to disable everything but the `light` theme:

```rust
                r#"<link rel="stylesheet" type="text/css" href="{}.css"{}>"#,
                Escape(&format!("{}{}{}", static_root_path, t, page.resource_suffix)),
                if t == "light" { "" } else { " disabled" }
```

And then in the theme switcher we simply remove the `disabled` attribute from the element before setting the `disabled` stylesheet property (confusing naming, right?):

```js
            themeSheet.removeAttribute('disabled');
            themeSheet.disabled = themeName !== newTheme;
```

This works well across most things I tested:

* Firefox with JS enabled (everything works as expected)
* Firefox with JS disabled (light theme is the only active stylesheet)
* Waterfox classic with JS enabled (everything works as expected)

Unfortunately Waterfox Classic with JS disabled is still broken because, well, it doesn't support the `disabled` attribute on links. Which is exactly what this PR was trying to remove.

I would guess that the set of people using Waterfox Classic is going to happen to intersect with the set of people that want to be able to use rustdoc with JS disabled, so this doesn't really accomplish anything.

The other idea I had was to only load the dark / ayu themes if JS is enabled (by adding them dynamically on page load). Should I try that out or nah?

> Maybe we could have light.css for <noscript> and turn it on with JS only otherwise?

@jyn514 good though, but unfortunately I don't think `<noscript>` can help us out here since there's no way for us to prevent the other theme stylesheets from getting loaded and applied to the page when JS is off.

**Comment by jyn514** [comment]
>  I don't think <noscript> can help us out here since there's no way for us to prevent the other theme stylesheets from getting loaded and applied to the page when JS is off.

Can we insert the other stylesheets with JS? So they don't show up at all in the HTML and only appear once the JS executes?

**Comment by jyn514** [comment]
and if that doesn't work I really think we should close this as wontfix, it's not worth the effort for fixing such a small minority of browsers.

**Comment by jyn514** [comment]
> > I don't think  can help us out here since there's no way for us to prevent the other theme stylesheets from getting loaded and applied to the page when JS is off.
> 
> Can we insert the other stylesheets with JS? So they don't show up at all in the HTML and only appear once the JS executes?

https://stackoverflow.com/a/577002/7669110 might possibly be what we want

**Comment by Cldfire** [comment]
> https://stackoverflow.com/a/577002/7669110 might possibly be what we want

I tried something similar to this and it looks like it will work. I don't have time to finish it right now though.

Will return and poke at it another night.

**Comment by Cldfire** [comment]
Branch [here](https://github.com/Cldfire/rust/tree/rustdoc-disable-stylesheet-js-add-stylesheets) that implements that idea (add stylesheets to page using JS in the head tag). Unfortunately this causes flashing on page load on my machine.

Unless anyone has any other bright ideas we are unfortunately going to have to close this PR :disappointed:

**Comment by GuillaumeGomez** [comment]
Well, at least now we know...

**Pull request closed without merging**
